src/types/redis_string.cc

```diff
@@ -530,4 +531,155 @@ rocksdb::Status String::CAD(const std::string &user_key, const std::string &valu
+  }
+  std::vector<uint32_t> dp(dp_size, 0);
+  auto lcs = [&](const uint32_t i, const uint32_t j) -> uint32_t & { return dp[i * (blen + 1) + j]; };
+
+  // Start building the LCS table.
+  for (uint32_t i = 1; i <= alen; i++) {
+    for (uint32_t j = 1; j <= blen; j++) {
+      if (a[i - 1] == b[j - 1]) {
+        // The len LCS (and the LCS itself) of two
+        // sequences with the same final character, is the
+        // LCS of the two sequences without the last char
+        // plus that last char.
+        lcs(i, j) = lcs(i - 1, j - 1) + 1;
+      } else {
+        // If the last character is different, take the longest
+        // between the LCS of the first string and the second
+        // minus the last char, and the reverse.
+        lcs(i, j) = std::max(lcs(i - 1, j), lcs(i, j - 1));
+      }
+    }
+  }
+
+  uint32_t idx = lcs(alen, blen);
+
+  // Only compute the length of LCS.
+  if (auto result = std::get_if<uint32_t>(rst)) {
+    *result = idx;
+    return rocksdb::Status::OK();
+  }
+
+  // Store the length of the LCS first if needed.
+  if (auto result = std::get_if<StringLCSIdxResult>(rst)) {
+    result->len = idx;
+  }
+
+  // Allocate when we need to compute the actual LCS string.
+  if (auto result = std::get_if<std::string>(rst)) {
+    result->resize(idx);
+  }
+
+  uint32_t i = alen;
+  uint32_t j = blen;
+  uint32_t arange_start = alen;  // alen signals that values are not set.
+  uint32_t arange_end = 0;
+  uint32_t brange_start = 0;
+  uint32_t brange_end = 0;
+  while (i > 0 && j > 0) {
+    bool emit_range = false;
+    if (a[i - 1] == b[j - 1]) {
+      // If there is a match, store the character if needed.
+      // And reduce the indexes to look for a new match.
+      if (auto result = std::get_if<std::string>(rst)) {
+        result->at(idx - 1) = a[i - 1];
+      }
+
+      // Track the current range.
+      if (arange_start == alen) {
+        arange_start = i - 1;
+        arange_end = i - 1;
+        brange_start = j - 1;
+        brange_end = j - 1;
+      }
+      // Let's see if we can extend the range backward since
+      // it is contiguous.
+      else if (arange_start == i && brange_start == j) {
+        arange_start--;
+        brange_start--;
+      } else {
+        emit_range = true;
+      }
+
+      // Emit the range if we matched with the first byte of
+      // one of the two strings. We'll exit the loop ASAP.
+      if (arange_start == 0 || brange_start == 0) {
+        emit_range = true;
+      }
+      idx--;
+      i--;
+      j--;
+    } else {
+      // Otherwise reduce i and j depending on the largest
+      // LCS between, to understand what direction we need to go.
+      uint32_t lcs1 = lcs(i - 1, j);
+      uint32_t lcs2 = lcs(i, j - 1);
+      if (lcs1 > lcs2)
+        i--;
+      else
+        j--;
+      if (arange_start != alen) emit_range = true;
+    }
+
+    // Emit the current range if needed.
+    uint32_t match_len = arange_end - arange_start + 1;
```

Comment:
put this in `emit_range`?

Reply:
I might not fully grasp your point. Are you suggesting that we should make `emit_range = arange_end - arange_start + 1`?

Reply:
```
// Emit the current range if needed.
- uint32_t match_len = arange_end - arange_start + 1;
if (emit_range) {
+   uint32_t match_len = arange_end - arange_start + 1;
```